### PR TITLE
GPU version of a 2-layer hierarchical softmax (ccw #2501)

### PR DESCRIPTION
**Pull request by adbrebs**
#2501

### CODE REVIEW

File: theano/sandbox/cuda/nnet.py [hunk 1]
```diff
@@ -1,5 +1,9 @@
+import theano
 from theano import Op, Apply
+from theano import tensor as T
 from theano.compat.six import StringIO
+from theano.sandbox.cuda.blocksparse import sparse_block_dot_SS
```
Comment (bouthilx) [inline theano/sandbox/cuda/nnet.py]: You should move this under "from theano.sandbox.cuda.basic_ops import as_cuda_ndarray_variable".

File: theano/sandbox/cuda/nnet.py [hunk 2]
```diff
@@ -721,3 +725,120 @@ def c_support_code_apply(self, node, nodename):
...
+    in sqrt(n_outputs) classes.
+    There are two softmax layers. The first predicts the class of the input x
+    while the second predicts the output of the input x in the predicted class.
+    More explanations can be found in the original paper
+    http://arxiv.org/abs/cs/0108006.
+    If target is specified, it will only compute the outputs of the
+    corresponding targets. Otherwise, if target is None, it will compute all
+    the outputs.
+
+    :type W1: a Theano shared variable of shape (number of features of the
```
Comment (bouthilx) [inline theano/sandbox/cuda/nnet.py]: I would describe parameters as [here](https://github.com/Theano/Theano/blob/master/theano/tensor/nnet/conv.py#L39). Shape is defined in ```:param name:``` and ```:type name:``` is simply ```symbolic 2D tensor``` (which holds for both TensorVariable and SharedVariable).

Comment (bouthilx) [inline theano/sandbox/cuda/nnet.py]: The same apply to all parameters descriptions

File: theano/sandbox/cuda/nnet.py [hunk 3]
```diff
@@ -721,3 +725,120 @@ def c_support_code_apply(self, node, nodename):
...
+
+    :note: If n_outputs is not a square number, it will still work but the
+        user has to specify the following shapes of its weight/bias matrices:
+        - number of outputs per class: n_outputs_per_class = ceil(sqrt(
+          n_outputs)
+        - number of classes: n_classes = ceil(n_outputs / n_outputs_per_class)
+
+    """
+
+    if theano.config.device != 'gpu':
```
Comment (bouthilx) [inline theano/sandbox/cuda/nnet.py]: It seems better to test [this way](https://github.com/Theano/Theano/blob/master/theano/sandbox/cuda/tests/test_nnet.py#L12)

File: theano/sandbox/cuda/nnet.py [hunk 4]
```diff
@@ -721,3 +725,120 @@ def c_support_code_apply(self, node, nodename):
...
+        raise Exception('The multiclass softmax only works on gpu.')
+
+    batch_size = x.shape[0]
+
+    n_classes, n_outputs_per_class = b2.get_value(True, True).shape
```
Comment (bouthilx) [inline theano/sandbox/cuda/nnet.py]: In conv2d case they decided to support any symbolic (4D) tensor. Maybe we should do so here too.

@nouiz @abergeron What do you think? SharedVariable is by far the most common use case and enable us to simplify the function signature. Do you think we should support TensorVariable anyway?

File: theano/sandbox/cuda/nnet.py (hunk 2 above)
Comment (bouthilx) [inline theano/sandbox/cuda/nnet.py]: Wait before we decide whether parameters can be any symbolic tensor or need to be SharedVariable

File: theano/sandbox/cuda/nnet.py [hunk 5]
```diff
@@ -721,3 +725,120 @@ def c_support_code_apply(self, node, nodename):
...
+
+        # Classes to which belong each target
+        target_classes = target // n_outputs_per_class
+
+        # Outputs to which belong each target inside a class
+        target_outputs_in_class = target % n_classes
+
+        # Second softmax that computes the output probabilities
+        # Adds a dimension so that sparse_block_dot_SS works properly
+        orig_shape = W2.get_value(True, True).shape
```
Comment (bouthilx) [inline theano/sandbox/cuda/nnet.py]: Have to make a decision here too.

**Comment by bouthilx** [comment]
Seems good. Only a few minor changes needed. Wait for Fred and/or Arnaud opinion.

**Comment by bouthilx** [comment]
Could you compute by hand (with numpy) the output of the hierarchical softmax given a fixed target?

**Comment by nouiz** [comment]
I let  @abergeron check it, as he know more the sparse block that is used
in this PR

File: theano/sandbox/cuda/nnet.py (hunk 4 above)
Comment (abergeron) [inline theano/sandbox/cuda/nnet.py]: How would restricting it to a shared variable simplify the function signature?

Comment (abergeron) [inline theano/sandbox/cuda/nnet.py]: Oh, I see, what you're doing.  But you can't count on the shape of a shared variable not changing.  It's not a constant.

File: theano/sandbox/cuda/nnet.py (hunk 3 above)
Comment (abergeron) [inline theano/sandbox/cuda/nnet.py]: I would just remove the test and force the params to the GPU with as_cudandarray_variable.  Just document that it only works on the GPU.

File: theano/sandbox/cuda/nnet.py [hunk 6]
```diff
@@ -721,3 +725,120 @@ def c_support_code_apply(self, node, nodename):
...
+    if target is None:
+        # Computes the probabilites of all the outputs
+
+        class_ids = T.arange(n_classes)
+
+        def _compute_output_probs(class_id):
+            # Second softmax that computes the probabilities of the outputs
+            # in a given class indexed by class_id
+            output_prob = softmax(
+                T.dot(x, W2[class_id, :, :]) + b2[class_id, :])
```
Comment (abergeron) [inline theano/sandbox/cuda/nnet.py]: This line looks exactly like a sparse_block_dot_SS.

File: theano/sandbox/cuda/nnet.py [hunk 7]
```diff
@@ -721,3 +725,120 @@ def c_support_code_apply(self, node, nodename):
...
+
+        # Second softmax that computes the output probabilities
+        # Adds a dimension so that sparse_block_dot_SS works properly
+        orig_shape = W2.get_value(True, True).shape
+        W2 = T.reshape(W2, (1,) + orig_shape)
+        output_probs = softmax(sparse_block_dot_SS(
+            W2, x[:, None, :], T.zeros((batch_size, 1), dtype='int64'), b2,
+            target_classes[:, None])[:, 0, :])
+        # Restores the original shape
+        W2 = T.reshape(W2, orig_shape)
```
Comment (abergeron) [inline theano/sandbox/cuda/nnet.py]: This is completely useless.

**Comment by abergeron** [comment]
From the travis tests it seems there is a problem importing blocksparse.  It would be a good idea to fix that also, but I'm not sure what the problem is exactly.

**Comment by abergeron** [comment]
I think it's a circular import between \__init__.py, nnet.py and blocksparse.py.  This means that you either can't import blocksparse from nnet or that you might have to import GpuElemwise directly in blocksparse.

File: theano/sandbox/cuda/nnet.py (hunk 3 above)
Comment (bouthilx) [inline theano/sandbox/cuda/nnet.py]: Makes sense, thanks!

File: theano/sandbox/cuda/nnet.py (hunk 4 above)
Comment (bouthilx) [inline theano/sandbox/cuda/nnet.py]: OK, we agreed to add n_classes and n_outputs_per_class to function arguments.

**Comment by bouthilx** [comment]
@abergeron is right. nnet imports blocksparse which imports opt which imports nnet... -_- Couldn't we load theano.sandbox.cuda as cuda with all namespace under it?

**Comment by nouiz** [comment]
You can do something like:

from theano.sandbox.cuda as cuda

then use

cuda.basic_ops....

Or anything else that is imported by default under cuda.

We use that at a few place to fix this type of problem.

File: theano/sandbox/cuda/nnet.py (hunk 6 above)
Comment (bouthilx) [inline theano/sandbox/cuda/nnet.py]: Looks more like a tensordot. The scan seems to do something like 

```python 
output_probs = (T.dot(x, W2) + b2) * class_probs[:,:,None]
output_probs = output_probs.reshape((batch_size, n_classes*n_outputs_per_class))[:,:n_outputs]
```

**Comment by adbrebs** [comment]
Thanks @bouthilx, @abergeron, @nouiz for your remarks. I have pushed some corrections. Please let me know if I did the right thing for the circular import.
I will also include the CPU version in this PR (in order to test GPU against CPU).

File: theano/sandbox/cuda/__init__.py [hunk 8]
```diff
@@ -297,6 +297,8 @@ def make_thunk(self, node, storage_map, compute_map, no_recycling):
             tensor3, tensor4)
     from .basic_ops import (host_from_gpu, gpu_from_host,
             as_cuda_array, as_cuda_ndarray_variable)
+    from . import blocksparse
```
Comment (abergeron) [inline theano/sandbox/cuda/__init__.py]: You should merge this line with the one below that also imports from `.`

**Comment by adbrebs** [comment]
@bouthilx, in which file shall we add the CPU version? 
Instead of having two functions, another possibility would be to have a single function with a device flag, which would be by default the flag defined by theano.config.device.

**Comment by adbrebs** [comment]
I think having a single function would have more sense because the prototype is the same for both cpu and gpu versions.

**Comment by adbrebs** [comment]
@bouthilx. I have pushed my last commits!

**Comment by bouthilx** [comment]
I discussed with @abergeron and a better way to handle both CPU and GPU would be to write a simple CPU version of the op ```SparseBlockGemvSS``` and add an optimization to swap CPU and GPU ops. A naive python implementation of ```perform``` method would be sufficient for now.

**Comment by nouiz** [comment]
What about doing a meta op without any implementation? Then you could have 2 optimizations: one to move it to the GPU. The second is after the optimization phase that move computation to the GPU, it would replace that op with the CPU or GPU graph.

This do not  need any perform implementation. What do you think of that? We are going in that direction for the structure of other part of Theano.

**Comment by bouthilx** [comment]
@nouiz You mean a meta op for ```SparseBlockGemvSS``` or for hierarchical softmax? We still need to write the code for CPU/GPU graph right? I never played with meta ops. Can you point to any doc about this?

**Comment by nouiz** [comment]
I mean a meta op for hierarchical softmax.

The optimization will replace that meta op with a theano graph that do the
right computation, but it allow to more easily have very different graph on
CPU and GPU.

There is no doc on this.

**Comment by abergeron** [comment]
The thing is, this will be used on the CPU, so we need a working implementation anyway.  I agree that the Meta opt is a better long-term strategy though so we may as well do that right now.

**Comment by nouiz** [comment]
I was understanding that the CPU and GPU implementation would be very
different graph. I think that is the case. But I was also under the
impression that the CPU graph already exist. If not, that is the first step
for sure.

**Comment by bouthilx** [comment]
OK I think I understand. Can you point any metaop we can base our work on? From what I understand, it will be almost an empty shell, and we'll write an optimizer function that will replace it with a graph we can wrap in some private ```_hierarchical_softmax_{cpu,gpu}``` functions like the ones @adbrebs already wrote.

**Comment by nouiz** [comment]
You seem to have understood it. There is no example, we are working on
doing the same for the convolution code, but it isn't merged and is
probably too complicated. But if you want to look, here is the PR:

https://github.com/Theano/Theano/pull/2665

**Comment by bouthilx** [comment]
@adbrebs @nouiz The squeleton of the meta op would be :
```python
class HierarchicalSoftmax(Op):
    # props = (properties...)

    def __init__(self, ?):
        # save arguments...

    def make_node(self, inputs... ):
        # test inputs...
        return Apply(self, inputs..., [output])

    def perform(self, node, inp, out_):
        raise NotImplementedError('HierarchicalSoftmax theano optimization failed')

    def grad(self, inp, grads):
        # return grad meta op?
```

The grad method bothers me. Our function output is a graph, so it's easy to compute the grad without adding any code, but now HierarchicalSoftmax.grad must return something. Is to possible to avoid having to write a grad meta op? The only solution I see would be return the meta-op itself with an additional flag saying we are looking for the grad during optimization, but that is somehow hacky.

**Comment by adbrebs** [comment]
Yes, I'm blocked because I don't know see any elegant way to:
- compute the gradient
- detect in the optimizer if the code runs on the gpu or on the cpu

**Pull request closed without merging**